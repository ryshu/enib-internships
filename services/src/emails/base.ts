import path from 'path';

import { email } from './config';

/**
 * @summary Method used to send welcom message to new user
 * @param {string} to Dest user
 * @returns {Promise<string>} Resolve: send message
 */
export function sendWelcom(to: string): Promise<string> {
    return email.send({
        template: path.join(__dirname, 'templates', 'welcome'),
        message: { to },
        locals: {},
    });
}
