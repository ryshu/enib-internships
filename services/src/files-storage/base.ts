import multer from 'multer';
import path from 'path';
import httpStatus from 'http-status-codes';

import { pdfFilter } from './filter';
import { APIError } from '../utils/error';

const baseDir = path.join(__dirname, '..', process.env.BASE_STORAGE_DIR);

export const multerPdfStorage = multer.diskStorage({
    destination(_req: any, _file, cb) {
        cb(null, baseDir);
    },
    filename(_req: any, file, cb) {
        if (file.originalname) {
            cb(
                null,
                file.originalname
                    .replace(/[^a-z0-9]/gi, '_')
                    .toLowerCase()
                    .replace(/__+/g, '-'),
            );
        } else {
            cb(new APIError('ID is required to upload a banner', httpStatus.BAD_REQUEST, 1), '');
        }
    },
});

export const uploadHandler = multer({
    storage: multerPdfStorage,
    fileFilter: pdfFilter,
    limits: {
        fileSize: 4000000, // 4 MB
        fields: 20,
        files: 1,
        parts: 20,
    },
});
