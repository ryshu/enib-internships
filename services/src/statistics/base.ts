export declare interface Statistics {
    internships: {
        total: number;
        suggested: number;
        waiting: number;
        availables: number;
        attributed: number;
        validated: number;
        archived: number;
    };

    students: number;
    mentors: number;

    propositions: number;
}

export declare interface CampaignStatistics {
    internships: {
        total: number;
        availables: number;
        attributed: number;
    };

    students: number;
    mentors: number;

    propositions: number;
    campaign: number;
}

export declare const enum INTERNSHIP_MODE {
    SUGGESTED = 'suggest',
    WAITING = 'waiting',
    AVAILABLE = 'available',
    IN_ATTRIBUTION = 'in_attribution',
    ATTRIBUTED = 'attributed',
    VALIDATED = 'validated',
    ARCHIVED = 'archived',
}

export const InternshipsMode: INTERNSHIP_MODE[] = [
    INTERNSHIP_MODE.SUGGESTED,
    INTERNSHIP_MODE.WAITING,
    INTERNSHIP_MODE.AVAILABLE,
    INTERNSHIP_MODE.ATTRIBUTED,
    INTERNSHIP_MODE.VALIDATED,
    INTERNSHIP_MODE.ARCHIVED,
];

export function isInternshipMode(str: string): str is INTERNSHIP_MODE {
    return InternshipsMode.includes(str as INTERNSHIP_MODE);
}
