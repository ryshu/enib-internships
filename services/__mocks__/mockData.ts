export function defaultBusiness() {
    const VALID_BUSINESS: IBusinessEntity = {
        name: 'test',
        country: 'France',
        city: 'Brest',
        postalCode: '29200',
        address: 'TEST',
    };
    return VALID_BUSINESS;
}

export function defaultFiles() {
    const VALID_FILE: IFileEntity = {
        name: 'fichier',
        size: 150,
        type: 'john.doe@enib.fr',
        path: 'S10',
    };
    return VALID_FILE;
}

export function defaultInternships() {
    const VALID_INTERNSHIP: IInternshipEntity = {
        subject: 'test',
        description: 'Stage',
        country: 'France',
        city: 'Brest',
        postalCode: '29280',
        address: 'TEST',
        isLanguageCourse: true,
        isValidated: false,
    };
    return VALID_INTERNSHIP;
}

export function defaultMentoringPropositions() {
    const VALID_MENTORING_PROPOSITION: IMentoringPropositionEntity = {
        comment: 'test',
    };
    return VALID_MENTORING_PROPOSITION;
}

export function defaultStudents() {
    const VALID_STUDENT: IStudentEntity = {
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@enib.fr',
        semester: 'S10',
    };
    return VALID_STUDENT;
}

export function defaultInternshipTypes() {
    const VALID_INTERNSHIP_TYPES: IInternshipTypeEntity = {
        label: 'test',
    };
    return VALID_INTERNSHIP_TYPES;
}
