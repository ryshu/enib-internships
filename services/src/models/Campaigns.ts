import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import MPS from './MentoringPropositions';
import Internships from './Internships';

class Campaigns extends Sequelize.Model implements ICampaignEntity {
    public static associations: {
        propositions: Sequelize.Association<Campaigns, MPS>;
        availbleInternships: Sequelize.Association<Campaigns, Internships>;
        validatedInternships: Sequelize.Association<Campaigns, Internships>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public name: string;
    public description: string;
    public startAt: number;
    public endAt: number;
    public semester: string;
    public maxProposition: number;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getPropositions: Sequelize.HasManyGetAssociationsMixin<MPS>;
    public addProposition: Sequelize.HasManyAddAssociationMixin<MPS, MPS['id']>;
    public createProposition: Sequelize.HasManyCreateAssociationMixin<IMentoringPropositionEntity>;
    public hasProposition: Sequelize.HasManyHasAssociationMixin<MPS, MPS['id']>;
    public countPropositions: Sequelize.HasManyCountAssociationsMixin;

    // AvailableInternships
    public getAvailableInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addAvailableInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createAvailableInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasAvailableInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countAvailableInternships: Sequelize.HasManyCountAssociationsMixin;

    // ValidatedInternships
    public getValidatedInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addValidatedInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createValidatedInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasValidatedInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countValidatedInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly propositions?: MPS[];
    public readonly availableInternships?: Internships[];
    public readonly validatedInternships?: Internships[];
}

Campaigns.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        name: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.STRING(),
            allowNull: false,
        },
        startAt: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            allowNull: true,
            defaultValue: null,
        },
        endAt: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            allowNull: true,
            defaultValue: null,
        },
        semester: {
            type: Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        maxProposition: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            defaultValue: 0,
            allowNull: true,
        },
    },
    {
        tableName: 'campaigns',
        sequelize: database,
    },
);

export default Campaigns;
