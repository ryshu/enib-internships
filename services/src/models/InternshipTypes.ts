import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import Internships from './Internships';

class InternshipTypes extends Sequelize.Model implements IInternshipTypeEntity {
    public static associations: {
        internships: Sequelize.Association<InternshipTypes, Internships>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public label: string;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly internships?: Internships[];
}

InternshipTypes.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        label: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: true,
        },
    },
    {
        tableName: 'internship-types',
        sequelize: database,
    },
);

export default InternshipTypes;
