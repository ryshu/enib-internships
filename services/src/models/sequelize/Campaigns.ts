import * as Sequelize from 'sequelize';

import database from '../../configs/instances/database';

import MPS from './MentoringPropositions';
import Mentors from './Mentors';
import InternshipTypes from './InternshipTypes';
import Internships from './Internships';

import {
    IInternshipEntity,
    IMentoringPropositionEntity,
    IInternshipTypeEntity,
} from '../../declarations';

class Campaigns extends Sequelize.Model {
    public static associations: {
        propositions: Sequelize.Association<Campaigns, MPS>;
        availbleInternships: Sequelize.Association<Campaigns, Internships>;
        validatedInternships: Sequelize.Association<Campaigns, Internships>;
        mentors: Sequelize.Association<Campaigns, Mentors>;
        category: Sequelize.Association<Campaigns, InternshipTypes>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public name: string;
    public description: string;

    public semester: string;
    public maxProposition: number;

    public isPublish: boolean;

    public startAt: number;
    public endAt: number;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    // Proposition
    public getPropositions: Sequelize.HasManyGetAssociationsMixin<MPS>;
    public addProposition: Sequelize.HasManyAddAssociationMixin<MPS, MPS['id']>;
    public createProposition: Sequelize.HasManyCreateAssociationMixin<IMentoringPropositionEntity>;
    public hasProposition: Sequelize.HasManyHasAssociationMixin<MPS, MPS['id']>;
    public countPropositions: Sequelize.HasManyCountAssociationsMixin;

    // Mentor
    public getMentors: Sequelize.BelongsToManyGetAssociationsMixin<Mentors>;
    public addMentor: Sequelize.BelongsToManyAddAssociationMixin<Mentors, Mentors['id']>;
    public hasMentor: Sequelize.BelongsToManyHasAssociationMixin<Mentors, Mentors['id']>;
    public countMentors: Sequelize.BelongsToManyCountAssociationsMixin;

    // Category
    public getCategory: Sequelize.BelongsToGetAssociationMixin<InternshipTypes>;
    public setCategory: Sequelize.BelongsToSetAssociationMixin<
        InternshipTypes,
        InternshipTypes['id']
    >;
    public createCategory: Sequelize.BelongsToCreateAssociationMixin<IInternshipTypeEntity>;

    // AvailableInternships
    public getAvailableInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addAvailableInternship: Sequelize.HasManyAddAssociationMixin<
        Internships,
        Internships['id']
    >;
    public createAvailableInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasAvailableInternship: Sequelize.HasManyHasAssociationMixin<
        Internships,
        Internships['id']
    >;
    public countAvailableInternships: Sequelize.HasManyCountAssociationsMixin;

    // ValidatedInternships
    public getValidatedInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addValidatedInternship: Sequelize.HasManyAddAssociationMixin<
        Internships,
        Internships['id']
    >;
    public createValidatedInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasValidatedInternship: Sequelize.HasManyHasAssociationMixin<
        Internships,
        Internships['id']
    >;
    public countValidatedInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly propositions?: MPS[];
    public readonly availableInternships?: Internships[];
    public readonly validatedInternships?: Internships[];
    public readonly mentors?: Mentors[];
    public readonly category: InternshipTypes;
}

Campaigns.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        name: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: false,
        },

        semester: {
            type: Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        maxProposition: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            defaultValue: 0,
            allowNull: true,
        },

        isPublish: {
            type: Sequelize.DataTypes.BOOLEAN,
            defaultValue: false,
        },

        startAt: {
            type: Sequelize.DataTypes.BIGINT,
            allowNull: true,
            defaultValue: null,
        },
        endAt: {
            type: Sequelize.DataTypes.BIGINT,
            allowNull: true,
            defaultValue: null,
        },
    },
    {
        tableName: 'campaigns',
        sequelize: database,
        defaultScope: {
            attributes: { exclude: ['categoryId'] },
        },
    },
);

export default Campaigns;
