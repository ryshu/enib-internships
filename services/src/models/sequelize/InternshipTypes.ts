import * as Sequelize from 'sequelize';

import database from '../../configs/instances/database';

import Internships from './Internships';
import Campaigns from './Campaigns';

import { ICampaignEntity, IInternshipEntity } from '../../declarations';

class InternshipTypes extends Sequelize.Model {
    public static associations: {
        internships: Sequelize.Association<InternshipTypes, Internships>;
        campaigns: Sequelize.Association<InternshipTypes, Campaigns>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public label: string;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    // Internships
    public getInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countInternships: Sequelize.HasManyCountAssociationsMixin;

    // Campaigns
    public getCampaigns: Sequelize.HasManyGetAssociationsMixin<Campaigns>;
    public addCampaign: Sequelize.HasManyAddAssociationMixin<Campaigns, Campaigns['id']>;
    public createCampaign: Sequelize.HasManyCreateAssociationMixin<ICampaignEntity>;
    public hasCampaign: Sequelize.HasManyHasAssociationMixin<Campaigns, Campaigns['id']>;
    public countCampaigns: Sequelize.HasManyCountAssociationsMixin;

    public readonly internships?: Internships[];
    public readonly campaigns?: Campaigns[];
}

InternshipTypes.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        label: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: true,
        },
    },
    {
        tableName: 'internship-types',
        sequelize: database,
        timestamps: true,
        paranoid: true,
    },
);

export default InternshipTypes;
