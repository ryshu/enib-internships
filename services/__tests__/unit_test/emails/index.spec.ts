import { sendWelcome, sendCampaignsCreate } from '../../../src/emails/base';

// Include database connection
import dbSetup from '../../../src/configs/setup/database';

import Campaigns from '../../../src/models/Campaigns';
import InternshipTypes from '../../../src/models/InternshipTypes';

import { defaultCampaigns, defaultInternshipTypes } from '../../../__mocks__/mockData';

jest.setTimeout(30000);

beforeAll((done) => {
    dbSetup.then(() => done()).catch((e) => done(e));
});

describe('Emails service', () => {
    it('sendWelcome_to_fr', async () => {
        const res = await sendWelcome('test@tester.fr');
        expect(res.envelope).toMatchSnapshot();
    });

    it('sendCampaignsCreate_to_fr', async () => {
        const VALID_CAMPAIGN = defaultCampaigns();
        const VALID_INTERNSHIP_TYPE = defaultInternshipTypes();

        const CREATED = await Campaigns.create(VALID_CAMPAIGN);
        const TYPE = await InternshipTypes.create(VALID_INTERNSHIP_TYPE);

        await CREATED.setCategory(TYPE.id);

        const res = await sendCampaignsCreate('test@tester.fr', CREATED);
        expect(res.envelope).toMatchSnapshot();
    });
});
