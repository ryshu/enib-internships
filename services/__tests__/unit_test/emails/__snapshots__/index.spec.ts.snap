// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emails service sendCampaignsCreate_to_fr 1`] = `
Object {
  "from": "no-reply@stages.enib.fr",
  "to": Array [
    "test@tester.fr",
  ],
}
`;

exports[`Emails service sendWelcome_to_fr 1`] = `
Object {
  "from": "no-reply@stages.enib.fr",
  "to": Array [
    "test@tester.fr",
  ],
}
`;
